integration/consume_integration_test.go

```diff
@@ -0,0 +1,273 @@
+
+	rw := newRemoteWrite(t)
+	defer rw.Close()
+
+	pulsarTopic := fmt.Sprintf("metrics-test-%s", randSeq(8))
+	pulsarURL := os.Getenv(envTestPulsarURL)
+
+	args := []string{
+		"consume",
+		// set pulsar url from environment
+		fmt.Sprintf(`--pulsar.url=%s`, pulsarURL),
+		fmt.Sprintf(`--pulsar.topic=%s`, pulsarTopic),
+		// use a random subscription name
+		fmt.Sprintf(`--pulsar.subscription=%s`, pulsarTopic),
+		// use a random port to listen for metrics
+		fmt.Sprintf(`--web.listen-address=%s`, metricHost),
+		// specifiy the remote write endpoint
+		fmt.Sprintf(`--remote-write.url=%s/api/v1/push`, rw.URL),
+	}
+
+	adapterCtx, adapterCancel := context.WithCancel(context.Background())
+	defer adapterCancel()
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		t.Logf("run adapter with args=%v", args)
+		err := app.Run(adapterCtx, args...)
+		assert.NoError(t, err)
+	}()
+
+	// connect to pulsar to produce some messages
+	pulsarClient, err := pulsar.NewClient(pulsar.ClientOptions{
+		URL: pulsarURL,
+	})
+	require.NoError(t, err)
+	defer pulsarClient.Close()
+
+	producer, err := pulsarClient.CreateProducer(pulsar.ProducerOptions{
+		Topic: pulsarTopic,
+	})
+	require.NoError(t, err)
+	defer producer.Close()
+
+	instanceLabels := instanceLabels(1, 501)
+	samplesEach := 4
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+
+		for i := samplesEach; i > 0; i-- {
+			produceBatch(t, ctx, producer, ti.tenantID, mpulsar.NewJSONSerializer(), instanceLabels)
+		}
+
+		// send an extra batch we don't check for, as for some reason sometimes
+		// a flake occurs where pulsar won't forward messages from the last
+		// batch. (according to pulsar's metrics, they have arrived correctly,
+		// but are not sent out to the subscription)
+		produceBatch(t, ctx, producer, ti.tenantID, mpulsar.NewJSONSerializer(), instanceLabels)
+
+		producer.Close()
+	}()
+
+	// receive all remote written samples and verify all samples per instance have arrived
+	var instancesToBeReceivedByTimestamp = make(instanceSampleMap)
+
+	for {
+		r := <-rw.reqCh
+		t.Logf("received request with %d series", len(r.req.Timeseries))
+
+		assert.Equal(t, ti.tenantID, mcontext.TenantIDFromContext(r.ctx), "expect tenantID to be propagated")
+
+		for _, series := range r.req.Timeseries {
+			instance, ok := labelGet(series.Labels, "instance")
+			if !ok {
+				panic("no instance label found")
+			}
+
+			for _, sample := range series.Samples {
+				toBeReceived, ok := instancesToBeReceivedByTimestamp[sample.Timestamp]
+				if !ok {
+					toBeReceived = newInstanceMap(instanceLabels)
+					instancesToBeReceivedByTimestamp[sample.Timestamp] = toBeReceived
+				}
+				delete(toBeReceived, instance)
+			}
+		}
+
+		if instancesToBeReceivedByTimestamp.finishedSamples() == samplesEach {
+			t.Logf("received %d samples per series", samplesEach)
```

Comment:
Maybe print the instance label here as well?

Reply:
This is actually for all instances, so only if we received a sample for all instance labels `finished samples` increases by one. Made that clearer in the log message:

```go
t.Logf("received %d samples per series for every instance", samplesEach)
```